File: ruby/run-length-encoding/run_length_encoding.rb
require 'ostruct'

class MappedData < Array
  def initialize
    @number_stack = []
  end

  def last_char?(char)
    last.char == char if last
  end

  def encode_char(char)
    last_char?(char) ? last.count += 1 : add_group(char)
  end

  def decode_char(char)
    if (0..9).to_a.map(&:to_s).include? char
      @number_stack << char
    else
      add_group(char, [@number_stack.join('').to_i, 1].max)
      @number_stack.clear
    end
  end

  def add_group(char, count = 1)
    push(CharacterGroup.new(char: char, count: count)) unless char.empty?
  end

  def encoded
    if empty?
      ''
    else
      map do |group|
        "#{group.count if group.count > 1}#{group.char}"
      end.join('')
    end
  end

  def decoded
    if empty?
      ''
    else
      map { |group| group.char * group.count }.join('')
    end
  end
end

class CharacterGroup < OpenStruct
end

class RunLengthEncoding
  def self.encode(original_data)
    mapped_data = MappedData.new
    original_data.chars.each { |char| mapped_data.encode_char(char) }
    mapped_data.encoded
  end

  def self.decode(encoded_data)
    mapped_data = MappedData.new
    encoded_data.chars.each { |char| mapped_data.decode_char(char) }
    mapped_data.decoded
  end
end

module BookKeeping
  VERSION = 3
end

File: javascript/run-length-encoding/run-length-encoding.js
export const encode = (str) => {
  if (str.length < 2) {return str}
  let encoded = '';
  let buffer = 1;
  for (let i = 0; i < str.length; i++) {
    let currentLetter = str[i];
    let nextLetter = str[i+1];
    if(currentLetter === nextLetter) {
      buffer++;
    } else if(buffer === 1) {
      encoded += currentLetter;
    } else {
      encoded += buffer + currentLetter;
      buffer = 1;
    }
  }
  return encoded;
};

export const decode = (str) => {
  if (str.length < 2) {return str}
  let decoded = '';
  let multiplier = '';
  for (let i = 0; i < str.length; i++) {
    let currentLetter = str[i];
    let nextLetter = str[i+1];
    if(currentLetter.match(/[0-9]/)) {
      multiplier += currentLetter;
    } else if(multiplier === '') {
      decoded += currentLetter;
    } else {
      decoded += currentLetter.repeat(parseInt(multiplier));
      multiplier = '';
    }
  }
  return decoded;
};
